src/core/DrumSynth.cpp

```diff
@@ -172,34 +171,26 @@ void DrumSynth::GetEnv(int env, const char* sec, const char* key, QString ini)
 
 float DrumSynth::waveform(float ph, int form)
 {
-	float w;
+	float w, p = ediv(ph, numbers::tau_v<float>);
 
 	switch (form)
 	{
-	case 0:
-		w = static_cast<float>(sin(fmod(ph, TwoPi)));
-		break; // sine
-	case 1:
-		w = static_cast<float>(fabs(2.0f * static_cast<float>(sin(fmod(0.5f * ph, TwoPi))) - 1.f));
-		break; // sine^2
-	case 2:
-		while (ph < TwoPi)
-		{
-			ph += TwoPi;
-		}
-		w = 0.6366197f * static_cast<float>(fmod(ph, TwoPi) - 1.f); // tri
-		if (w > 1.f)
-		{
-			w = 2.f - w;
-		}
+	case 0: // sine
+		w = std::sin(p);
+		break;
+	case 1: // sine^2
+		w = std::abs(2.f * std::sin(.5f * p)) - 1.f;
+		break;
+	case 2: // triangle
+		w = p * 2.f * numbers::inv_pi_v<float>;
+		if (w > 1.f) { w = 2.f - w; }
+		break;
+	case 3: // sawtooth
+		w = p * numbers::inv_pi_v<float> - 1.f;
```

Comment:
```suggestion
		w = (ph - (2 * numbers::inv_pi_v<float>) * static_cast<int>(ph / (2 * numbers::inv_pi_v<float>))) / numbers::inv_pi_v<float> - 1;
```
I think the previous equation was pretty good.

Reply:
I'm confused; the previous equation is not what is suggested here?
![image](https://github.com/user-attachments/assets/f9ee7dff-7dd7-4a39-8b81-145756557e46)

Reply:
Updated graph with new code changes:
![image](https://github.com/user-attachments/assets/8a041f9b-eafb-4570-912c-4dd09c0f77d3)

Reply:
Looks like this was my mistake copy-pasting `numbers::inv_pi_v<float>` instead of `numbers::pi_v<float>` again. Fixed.